value.go

```diff
@@ -317,7 +350,43 @@ type valueSimpleObject struct {
 	valueObjectBase
 	upValues bindingFrame
 	fields   valueSimpleObjectFieldMap
-	asserts  []ast.Node
+	asserts  []unboundField
+}
+
+func checkAssertionsHelper(e *evaluator, obj valueObject, curr valueObject, superDepth int) error {
+	switch curr := curr.(type) {
+	case *valueExtendedObject:
+		err := checkAssertionsHelper(e, obj, curr.right, superDepth)
+		if err != nil {
+			return err
+		}
+		err = checkAssertionsHelper(e, obj, curr.left, superDepth+curr.right.inheritanceSize())
+		if err != nil {
+			return err
+		}
+		return nil
+	case *valueSimpleObject:
+		for _, assert := range curr.asserts {
+			_, err := e.evaluate(assert.bindToObject(selfBinding{self: obj, superDepth: superDepth}, curr.upValues))
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	default:
+		panic(fmt.Sprintf("Unknown object type %#v", obj))
+	}
+}
+
+func checkAssertions(e *evaluator, obj valueObject) error {
+	if !obj.assertionsChecked() {
+		// Assertions may refer to the object that will normally
```

Comment:
I don't think I ever proved that this was sound.  It seems OK though :)

Reply:
I think it is.  Each object either passes the assertion or not. So the only thing that matters for the final result (i.e. error or no error) is the set of objects to check. It doesn't matter when we check them. We check them as soon as we can, b/c that is the most predictable, obvious thing to the user.

This would get more complicated if we added some sort of error handling.